crates/services/gas_price_service/src/v1/da_source_adapter/block_committer_costs.rs

```diff
@@ -6,57 +9,334 @@ use crate::v1::da_source_adapter::{
+        &self,
+        range: core::ops::Range<u64>,
+    ) -> DaBlockCostsResult<Vec<Option<RawDaBlockCosts>>> {
+        let response = self
+            .client
+            .get(&format!("{}/{}-{}", self.url, range.start, range.end))
+            .send()
+            .await?
+            .json::<Vec<RawDaBlockCosts>>()
+            .await?;
+        Ok(response.into_iter().map(Some).collect())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    struct MockBlockCommitterApi {
+        value: Option<RawDaBlockCosts>,
+    }
+
+    impl MockBlockCommitterApi {
+        fn new(value: Option<RawDaBlockCosts>) -> Self {
+            Self { value }
+        }
+    }
+
+    #[async_trait::async_trait]
+    impl BlockCommitterApi for MockBlockCommitterApi {
+        async fn get_latest_costs(&self) -> DaBlockCostsResult<Option<RawDaBlockCosts>> {
+            Ok(self.value.clone())
+        }
+        async fn get_costs_by_seqno(
+            &self,
+            seq_no: u64,
+        ) -> DaBlockCostsResult<Option<RawDaBlockCosts>> {
+            // arbitrary logic to generate a new value
+            let mut value = self.value.clone();
+            if let Some(value) = &mut value {
+                value.sequence_number = seq_no;
+                value.blocks_range =
+                    value.blocks_range.end * seq_no..value.blocks_range.end * seq_no + 10;
+                value.da_block_height = value.da_block_height + (seq_no + 1).into();
+                value.total_cost += 1;
+                value.total_size_bytes += 1;
+            }
+            Ok(value)
+        }
+        async fn get_bundles_by_range(
+            &self,
+            _: core::ops::Range<u64>,
+        ) -> DaBlockCostsResult<Vec<Option<RawDaBlockCosts>>> {
+            Ok(vec![self.value.clone()])
+        }
+    }
+
+    fn test_da_block_costs() -> RawDaBlockCosts {
+        RawDaBlockCosts {
+            sequence_number: 1,
+            blocks_range: 0..10,
+            da_block_height: 1u64.into(),
+            total_cost: 1,
+            total_size_bytes: 1,
+        }
+    }
+
+    #[tokio::test]
+    async fn request_da_block_cost__when_last_value_is_none__then_get_latest_costs_is_called(
+    ) {
+        // given
+        let da_block_costs = test_da_block_costs();
+        let mock_block_committer =
+            MockBlockCommitterApi::new(Some(da_block_costs.clone()));
+        let mut block_committer =
+            BlockCommitterDaBlockCosts::new(mock_block_committer, None);
+
+        // when
+        let block_committer_da_block_costs =
+            block_committer.request_da_block_cost().await.unwrap();
+
+        // then
+        assert_eq!(
+            block_committer_da_block_costs,
+            DaBlockCosts {
+                l2_block_range: da_block_costs.blocks_range.clone(),
+                blob_size_bytes: da_block_costs.total_size_bytes,
+                blob_cost_wei: da_block_costs.total_cost,
+            }
```

Comment:
nit: rename `block_committer_da_block_costs` to `actual` and the `DaBlockCosts` to `expected`.

Reply:
addressed in 9ea0a9cdaa